---
title: "Share Seq Exploration"
author: "Aidan McLoughlin"
date: "`r format(Sys.time(), '%B %d, %Y')`"

header-includes:
   - \usepackage{float}
   - \usepackage{hyperref}
   - \usepackage{tcolorbox}
   - \hypersetup{colorlinks, 
                 urlcolor = blue,
                 linkcolor = black, 
                 citecolor = black}
                 
output: html_document                 
---


```{r libraries}
library(tidyverse)
library(data.table)
library(here)
library(Matrix)
library(GenomicRanges)
library(Seurat)
library(Signac)
# library(EnsDb.Hsapiens.v86)
library(EnsDb.Mmusculus.v79) # For mouse brain


`%p%` <- function(x,y) paste0(x,y)
```



```{r setup-dirs}

file <- here("data/EmbryonicMouseBrain_10x_multiomics/e18_mouse_brain_fresh/" %p%  
               "e18_mouse_brain_fresh_5k_filtered_feature_bc_matrix.h5")

inputdata <- Read10X_h5(file)

# Genome matrix has multiple modalities, returning a list of matrices for this genome
dir.out <- here("output")

file <- "inputdata.rda"
path <- paste(dir.out, file, sep="/")
save(inputdata, file = path)
```

```{r setup}
rna_counts <- inputdata$`Gene Expression`
atac_counts <- inputdata$Peaks

e18 <- CreateSeuratObject(counts = rna_counts)
e18[["percent.mt"]] <- PercentageFeatureSet(e18, pattern = "^MT-")
grange.counts <- StringToGRanges(rownames(atac_counts), sep = c(":", "-"))
grange.use <- seqnames(grange.counts) %in% standardChromosomes(grange.counts)
atac_counts <- atac_counts[as.vector(grange.use), ]
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)
seqlevelsStyle(annotations) <- 'UCSC'
genome(annotations) <- "mm10"
```

```{r add-atac-frags}


frag.path <- here("data/EmbryonicMouseBrain_10x_multiomics/e18_mouse_brain_fresh/" %p%  
               "e18_mouse_brain_fresh_5k_atac_fragments.tsv.gz")

chrom_assay <- CreateChromatinAssay(
  counts = atac_counts,
  sep = c(":", "-"),
  genome = 'mm10',
  fragments = frag.path,
  min.cells = 10,
  annotation = annotations
)
e18[["ATAC"]] <- chrom_assay
```



```{basic qc}
## perform basic qc based on the number of detected molecules for each modality
## as well as mitochondrial percentage
VlnPlot(e18, features = c("nCount_ATAC", "nCount_RNA","percent.mt"), ncol = 3,
        log = TRUE, pt.size = 0) + NoLegend()
e18 <- subset(
  x = e18,
  subset = nCount_ATAC < 7e4 &
    nCount_ATAC > 5e3 &
    nCount_RNA < 25000 &
    nCount_RNA > 1000 &
    percent.mt < 20
)
e18
# An object of class Seurat 
# 170928 features across 3867 samples within 2 assays 
# Active assay: RNA (32285 features, 0 variable features)
# 1 other assay present: ATAC
```

```{r-umap}
## perform pre-processing and dimensional reduction on both assays independently
## using standard approaches for rna and atac-seq data.
DefaultAssay(e18) <- "RNA"
e18 <- SCTransform(e18, verbose = FALSE) %>%
  RunPCA() %>%
  RunUMAP(dims = 1:50, reduction.name = 'umap.rna', reduction.key = 'rnaUMAP_')
DefaultAssay(e18) <- "ATAC"
e18 <- FindTopFeatures(e18, min.cutoff = 5)
e18 <- RunTFIDF(e18)
e18 <- RunSVD(e18)
e18 <- RunUMAP(e18, reduction = 'lsi', dims = 2:50,
                reduction.name = "umap.atac",
                reduction.key = "atacUMAP_")


## calculate a wnn graph, representing a weighted combination of rna and atac-seq modalities
## use this graph for umap visualization and clustering
brain <- 
  FindMultiModalNeighbors(brain, 
                          reduction.list = list("pca", "lsi"), 
                          dims.list = list(1:50, 2:50))
```





```{gene-activity}


