---
title: "Share Seq Exploration"
author: "Aidan McLoughlin"
date: "`r format(Sys.time(), '%B %d, %Y')`"

header-includes:
   - \usepackage{float}
   - \usepackage{hyperref}
   - \usepackage{tcolorbox}
   - \hypersetup{colorlinks, 
                 urlcolor = blue,
                 linkcolor = black, 
                 citecolor = black}
                 
output: html_document                 
---


```{r libraries}
library(tidyverse)
library(data.table)
library(here)
library(Matrix)
library(GenomicRanges)
library(Seurat)
library(Signac)
# library(EnsDb.Hsapiens.v86)
library(EnsDb.Mmusculus.v79) # For mouse brain
```



```{r share-seq-brain-processing-1}

ct <- fread(here('data/share_seq/brain/celltype_brain.txt'))

ct <- ct %>% 
  mutate_at(vars(atac.bc, rna.bc), 
            ~ gsub("[[:punct:]]", "", .))

## Two cells have the same ATAC primer
# ct <- ct %>% 
#   mutate(atac_no_primer = str_sub(rna.bc, 1L, -7))
# 
# frag.bed <- frag.bed %>% 
#   mutate(V4_no_primer = str_sub(V4, 1L, -7))
# 
# setdiff(frag.bed$V4_no_primer, ct$atac_no_primer) %>% length()

```

```{r}

# first column: atac barcode
# second column: rna barcode
# third column: annotated cell type from the paper

rna.count <-
  fread(here('data/share_seq/brain/GSM4156610_brain.rna.counts.txt.gz'))

# columns are barcodes, rows are genes
rna.count <- rna.count %>% 
  as.data.frame()

row.names(rna.count) <- rna.count$gene

rna.count <-
  rna.count %>% dplyr::select(-gene)

names(rna.count) <- gsub("[[:punct:]]", "", names(rna.count))
#names(rna.count) <- gsub("\\,", "\\.", names(rna.count))
  
```

```{r share-seq-brain-processing-2}
# match the rna.count with the cell type labels
any(!is.na(match(ct$rna.bc, colnames(rna.count))))


match(ct$rna.bc, colnames(rna.count)) %>% as.data.frame() %>% summary(.)

rna.count<-
  rna.count[, match(ct$rna.bc, colnames(rna.count))]


```




```{r share-seq-brain-processing-3}
# The barcode file is the same as the annotation ct; omitted
peak.barcodes <- scan(here('data/share_seq/brain/GSM4156599_brain.barcodes.txt.gz'), 
                      what = "")
peak.barcodes <- gsub("[[:punct:]]", "", peak.barcodes)

all(ct$rna.bc == peak.barcodes)
rm(peak.barcodes)
```

```{r share-seq-brain-processing-4}
# Read in the peak matrix: stored as Market matrix
# https://atlas.gs.washington.edu/mouse-atac/docs/

peak.count <- 
  readMM(here("data/share_seq/brain/GSM4156599_brain.counts.txt.gz"))




## Vector with which we place genomic locations of sequences 
##    and will use GRanges function to annotate
peak.bed <- 
  read.delim(here('data/share_seq/brain/GSM4156599_brain.peaks.bed.gz'), 
             header=FALSE)


peak.granges <- GRanges(seqnames = peak.bed$V1, 
                        ranges = IRanges(st  = peak.bed$V2, 
                                         end = peak.bed$V3))
rm(peak.bed)

rownames(peak.count) <- paste(peak.granges)
colnames(peak.count) <- ct$rna.bc # we will use the rna barcode from here and on
#colnames(peak.count) <- ct$atac.bc 


dim(ct) # barcode and cell type
dim(peak.count) # peak count
length(peak.granges) # GRanges for peaks
dim(rna.count) # rna count



```


```{r mitochondria-expression}

# # None is returned with the command below
#   (IE:  searching for genes beginning with "mt-" in name)
# rownames(rna.count)[grep('^mt-', rownames(rna.count), ignore.case = TRUE)]


# # Try another approach https://www.michaelchimenti.com/2019/03/calculate-mitochondrial-for-mouse-scrna-seq/

## Common preprocessing step is to filter out cells with certain percent Mitochondrial gene expression

mouse_mito <- 
  as_tibble(
    read.csv(here("data/share_seq/brain/Mouse.MitoCarta2.0.txt"), 
             header = TRUE, sep='\t'))

mouse_mito <- mouse_mito %>% 
  dplyr::select(c(Symbol, MCARTA2.0_score)) %>% 
  dplyr::slice(1:100)

mito.genes <- as.character(mouse_mito$Symbol)
mito.genes <- mito.genes[mito.genes %in% rownames(rna.count)]
rownames(rna.count)[match(mito.genes, rownames(rna.count))] <- 
  paste('MT-', mito.genes, sep='')
# # The results do not look good. The list may not be as accurate.
```


```{r finalize-rna-data}
# Create Seurat object
brain <- CreateSeuratObject(counts = rna.count)
brain[["percent.mt"]] <- PercentageFeatureSet(brain, pattern = "^MT-")
```


```{r finalize-atac-data}
# Now add in the ATAC-seq data
# we'll only use peaks in standard chromosomes
grange.use <- 
  Signac::seqnames(peak.granges) %in% standardChromosomes(peak.granges)


peak.count <- 
  peak.count[as.vector(grange.use), ]
peak.granges <- 
  peak.granges[as.vector(grange.use)]

rm(grange.use)

annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)

## move to UCSC annotation style
seqlevelsStyle(annotations) <- 'UCSC'

# Mus musculus (house mouse) genome assembly GRCm38 (mm10)
genome(annotations) <- "mm10"


```

```{r frag-file}
## Unique ATAC fragments from the reads
frag.file <- here("data/share_seq/brain/GSM4156599_brain.atac.fragments.bed.gz")

# frag.bed <-
#   read.delim(here('data/share_seq/brain/GSM4156599_brain.atac.fragments.bed.gz'),
#              header=FALSE)
# 
# 
# frag.cells <-
#   unique(frag.bed$V4)
# 
# length(setdiff(ct$atac.bc, frag.cells))

## Example file:
#library(hdf5r)
# counts <- Read10X_h5(filename = here("data/sample_files/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5"))
# metadata <- read.csv(
#   file = here("data/sample_files/atac_v1_pbmc_10k_singlecell.csv"),
#   header = TRUE,
#   row.names = 1
# )



### ATM question: Does this function generate the normalized counts for you?
chrom_assay <- CreateChromatinAssay(
  counts = peak.count,
  sep = c(":", "-"),
  genome = 'mm10',
  #fragments = frag.file,
  ## does this need to be 10?
  min.cells = 0,
  annotation = annotations
)
```

```{r combine-rna-atac}
brain[["ATAC"]] <- chrom_assay
brain@assays$RNA
brain@assays$ATAC
```


```{r qc-exploration-1}
## perform basic qc based on the number of detected molecules for each modality
## as well as mitochondrial percentage
Seurat::VlnPlot(brain, 
                features = c("nCount_ATAC", "nCount_RNA","percent.mt"), 
                ncol = 3, 
                log = TRUE, 
                pt.size = 0) + NoLegend()

```

```{r qc-exploration-2}
brain <- subset(
  x = brain,
  subset = nCount_ATAC < 3000 &
    nCount_ATAC > 100 &
    nCount_RNA < 25000 &
    nCount_RNA > 500 &
    percent.mt < 20
)
VlnPlot(brain, 
        features = c("nCount_ATAC", "nCount_RNA","percent.mt"), 
        ncol = 3,
        ## just changes visual scaling of y-axis
        log = TRUE, 
        pt.size = 0) + NoLegend()

```



```{r pre-processing-1, message=FALSE, warning=FALSE}
## perform pre-processing and dimensional reduction on both assays independently
## using standard approaches for rna and atac-seq data
DefaultAssay(brain) <- "RNA"
## ATM: why dims 1-50 as argument here?
brain <- SCTransform(brain, verbose = FALSE) %>%
  RunPCA() %>%
  RunUMAP(dims = 1:50, 
          reduction.name = 'umap.rna', 
          reduction.key = 'rnaUMAP_')
```


```{r pre-processing-2, warning=FALSE, message=FALSE}
DefaultAssay(brain) <- "ATAC"
brain <- FindTopFeatures(brain, min.cutoff = 5)
brain <- RunTFIDF(brain)
brain <- RunSVD(brain)

## why dims 2-50 here? 
# explanation of current scATAC dim reduction considerations 
##  and Latent Semantic Indexing (LSI) approach (vs PCA): 
##  http://andrewjohnhill.com/blog/2019/05/06/dimensionality-reduction-for-scatac-data/
brain <- RunUMAP(brain, 
                 reduction = 'lsi', 
                 dims = 1:50,
                 reduction.name = "umap.atac",
                 reduction.key = "atacUMAP_")
```

```{r wnn-graph}
## calculate a wnn graph, representing a weighted combination of rna and atac-seq modalities
## use this graph for umap visualization and clustering
brain <- 
  FindMultiModalNeighbors(brain, 
                          reduction.list = list("pca", "lsi"), 
                          dims.list = list(1:50, 2:50))

```
```{r umap-on-wnn}
brain <- RunUMAP(brain, nn.name = "weighted.nn", 
                 reduction.name = "wnn.umap", 
                 reduction.key = "wnnUMAP_")

### ATM: they are using the SLM algorithm here.
brain <- FindClusters(brain, 
                      graph.name = "wsnn", 
                      algorithm = 3, verbose = FALSE)
```

```{r plotting-of-data}
load(here("data/share_seq/brain/brain.rda"))
table(brain$seurat_clusters)

p1 <- DimPlot(brain, 
              reduction = "umap.rna", 
              group.by = "seurat_clusters", label = TRUE, 
              label.size = 2.5, repel = TRUE) + 
  ggtitle("RNA")

p2 <- DimPlot(brain, 
              reduction = "umap.atac", 
              group.by = "seurat_clusters", label = TRUE, 
              label.size = 2.5, repel = TRUE) + 
  ggtitle("ATAC")

p3 <- DimPlot(brain, 
              reduction = "wnn.umap", 
              group.by = "seurat_clusters", 
              label = TRUE, 
              label.size = 2.5, repel = TRUE) + 
  ggtitle("WNN")


## add annotated cell types
barcodes.brain <- 
  row.names(brain@meta.data) %>% 
  data.frame()
names(barcodes.brain) <- "barcode"

celltype_df <- ct %>% 
  dplyr::select(barcode = rna.bc, celltype)

barcodes.brain <- 
  barcodes.brain %>% 
  left_join(celltype_df)



brain@meta.data <-cbind(brain@meta.data, celltype = barcodes.brain$celltype)
  
  

p4 <- DimPlot(brain, 
              reduction = "wnn.umap", 
              group.by = "celltype", 
              label = TRUE, 
              label.size = 2.5, repel = TRUE) + 
  ggtitle("WNN")

p1 + p2 + p3 & NoLegend() & 
  theme(plot.title = element_text(hjust = 0.5))


p3 + p4 & NoLegend()

#save(brain, file='brain.rda')
```

```{r gene-activity-matrix}
# We will now get the gene activity matrix
# https://satijalab.org/signac/articles/brain_vignette.html
DefaultAssay(brain) <- "ATAC"

## Still need to figure out the issue with the fragment file
# gene.activities <- GeneActivity(brain) # This would not run because it requires fragment file to be loaded.

# Below is a short-cut but is not very precise: we only aggregate reads in the peak regions
# But there are reads off peak regions but within gene bodies that are defined
annotation <- Annotation(object = brain)
transcripts <- Signac:::CollapseToLongestTranscript(ranges = annotation)
transcripts <- transcripts[transcripts$gene_biotype == "protein_coding"]
transcripts <- Extend(x = transcripts, upstream = 2000, 
                      downstream = 0)
gene.activities=matrix(nrow=length(transcripts), ncol=ncol(brain))
rownames(gene.activities)=transcripts$gene_name
colnames(gene.activities)=colnames(brain)
peak.count=brain@assays$ATAC@counts
peak.granges=brain@assays$ATAC@ranges
for(i in 1:nrow(gene.activities)){
  if(i %%200 ==0) cat(i,' ')
  peak.index=which(countOverlaps(peak.granges, transcripts[i])>0)
  gene.activities[i,]=apply(peak.count[peak.index,, drop=FALSE],2, sum)
}
gene.activities=gene.activities[apply(gene.activities,1,sum)>10,]
gene.activities=gene.activities[!duplicated(rownames(gene.activities)),] # remove duplicate genes
gene.activities=gene.activities[-which(rownames(gene.activities)==''),]

brain[['ACTIVITY']] <- CreateAssayObject(counts = gene.activities)
## NOTE: WE WILL STILL NEED TO LOOK INTO THE FRAGMENT FILE;
## This calculation is not accurate since we only focus on the peak regions

# Perform normalization/scaling of the gene activity matrix
DefaultAssay(brain) <- "ACTIVITY"
brain <- NormalizeData(brain)
brain <- FindVariableFeatures(brain)
all.genes <- rownames(brain)
brain <- ScaleData(brain, features = all.genes)
brain@assays$ACTIVITY@scale.data[1:4,1:4]
#save(brain, file='brain.rda')




```




```{r gene-ontology}
library(R.utils)


gunzip(filename = here("data/gene_ontology/gene_association.mgi.gz"), 
       overwrite = TRUE, remove = FALSE)


brain_go <- fread(here('data/gene_ontology/GO_term_summary.txt'))


```



